### PR TITLE
Add namespace k8s tagger

### PR DESCRIPTION
**Pull request by chaitanyaphalak**
**Description:** Add feature to fetch namespace metadata, labels and annotations

**Link to tracking Issue:** Feature request - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3430 and use-case - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3031

**Testing:** 

**Documentation:**

### CODE REVIEW

**Comment by owais** [comment]
I think we should re-think the config changes this introduces. Instead of a new `extract_namespaces` section, we can extend the existing `extract` section. Some ideas,

1. Use path like keys under extract to tell the exporter to extract from something other than the pod. 
```
    extract:
      namespace/annotations:
      - key: splunk.com/index
      - key: splunk.com/sourcetype
      - key: splunk.com/exclude
```

This works well if we always want to support extacting data from objects that are related to a pod. In the example above, w'd extract the annotations from the namespace that the pod is in. In future we could have `deployment/*`, `daemonset/*`, `service/*` or any other type that could be related to the pod which generated the span.

2. Add a `from: ` attribute to every item to specify where to extract the item from. 
```
    extract:
      annotations:
      - key: splunk.com/index
        from: namespace
      - key: splunk.com/sourcetype
        from: namespace
      - key: splunk.com/exclude # gets extracted from pod by default
```

the new `from` attribute can default to `pod` and as a result needs to be mentioned only when extracting from something other than a pod.

I feel this allows for a much nicer config pattern that can be extended in future as we add support for newer types. Open to other ideas as well.

**Comment by chaitanyaphalak** [comment]
@owais 
I think the first option feels easy to use from a config and backward compatability point of view.

Should we follow the same pattern for metadata and labels?

    extract:
      namespace/labels:
        - tag_name: l1 # extracts value of label with key `label1` and inserts it as a tag with key `l1`
           key: label1      
    namespace/metadata:
        # extract the following well-known metadata fields
        - namespaceName
        - namespaceUID
        - namespaceCluster
        - namespaceStartTime

**Comment by owais** [comment]
> Should we follow the same pattern for metadata and labels?

Yes. I used annotations just as an example. We should support all feature we can/want for all k8s resources we support. 

I'd wait to see what others think about these config options before going ahead with the implementation. Tagging some people who might be interested in this.

@dmitryax @pmm-sumo @ericmustin @tigrannajaryan @pmatyjasek-sumo @dashpole

**Comment by dmitryax** [comment]
I like both options that Owais suggested, the option **2** looks probably a bit better to me. 

Having a default value `pod` for the additional `from` key looks more natural than assuming that `annotations` == `pod/annotations`.

**Comment by owais** [comment]
@dmitryax agree. It also makes it easier to deal with config if we don't have to split and compare strings.

**Comment by pmatyjasek-sumo** [comment]
I think the option 2 looks better. It is inline with for example `pod_association` convention

**Comment by tigrannajaryan** [comment]
What do we do with the additional well-know fields? Do they go under the existing "metadata" setting? e.g.:
```
    extract:
      metadata:
        # extract the following well-known metadata fields
        # these are pod fields:
        - podName
        - podUID
        - deployment
        - cluster
        - namespace
        - node
        - startTime
        # these are namespace fields:
        - namespaceName # is this the same as "namespace" above?
        - namespaceUID
        - namespaceCluster
        - namespaceStartTime
```
Do we believe that this single list of fields is sufficient for future needs? Do we expect field name clashes that make it impossible to disambiguate which entity the metadata should be extracted from? For example is "startTime" the pod start time or node start time (hypothetically speaking)?

**Comment by ericmustin** [comment]
don't have super strong opinions here, but #2 seems a bit cleaner generally speaking

**Comment by pmm-sumo** [comment]
No strong opinion here either. Option 2 seems cleaner but option 1 is less verbose. 

> Does extracted data get added as a Resource attribute or as a Span/LogRecord attribute?

I think that in principle, this processor should impact Resource-level attributes only and it seems [it wasn't changed here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3384/files#diff-d980cc2893cf804127a14a1a6d9a552a69a43bc76400dffdd155bb385b31c040R122-R129). If attributes are on a record-level, they can be extracted level up using [groupbyattrsprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/groupbyattrsprocessor)

**Comment by chaitanyaphalak** [comment]
@tigrannajaryan These are added as resource attributes. "startTime" is distinct for all object types like pods, namespaces etc, so they all need to distinct in some way. Also "namespace" and "namespaceName" are the same but extracted from different objects, pods and namespace respectively.

**Comment by tigrannajaryan** [comment]
Should we use Otel-compliant names for fields that we extract? So for example instead of:
```yaml
    extract:
      metadata:
        # extract the following well-known metadata fields
        # these are pod fields:
        - podName
        - podUID
        - deployment
        - cluster
        - namespace
        - node
        - startTime
```
we do:
```yaml
    extract:
      metadata:
        # extract the following well-known metadata fields
        # these are pod fields:
        - k8s.pod.name # this comes from a pod
        - k8s.pod.uid
        - k8s.deployment.name # where does this come from? from pod or the associated deployment? Or it is the same thing?
        - k8s.namespace.name # this comes from a namespace
        - k8s.pod.start_time # pod's start time.
        - k8s.node.start_time # node's start time.
        - startTime # alias for k8s.pod.start_time
        - etc ...
```
This way it is more apparent what is it that we are populating. This way we can also disambiguate fields which may be present in more than one entity. Thoughts?

I understand this means changing already existing field names, but we can do it in a backward compatible way by supporting  old names as aliases for the fields that we start supporting new canonical Otel names.

**Comment by chaitanyaphalak** [comment]
@tigrannajaryan that actually makes it really easy for the users to understand what they want and configure appropriately.

**Comment by owais** [comment]
@tigrannajaryan makes sense for the "metadata" section which is static list of well-known fields. One implication of this is that if we need support for new fields, we'll first need to go and update the semantic conventions (which is probably a good thing anyway). 

But I think this should be done in a separate work PR.

@chaitanyaphalak this reminds me that `metadata` section is list of well-known fields and the processor know exactly how and where to fetch these values from  ([ref](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sprocessor/options.go#L80-L98)). As such, I don't think it makes sense for users to specify a `from: ` option to metadata fields as the processor already knows how to extract them. It still makes sense for Annotations and Labels though. 

I think we should split this into multiple PRs:

1. First PR updates the field names to what Tigran suggested. Let's not add any new fields. Only rename.
2. (if applicable) PR adds new well-known fields that might possibly require us to fetch and watch namespace resources. 
3. Add `from: ` support to annotations and labels so they can be extracted from namespaces in addition to pods. 

WDYT?

**Comment by owais** [comment]
@chaitanyaphalak One more thing. If you add support for new metadata fields, try to restrict them in a way so they can be extracted from the pod spec and won't require watching resources other than pods. If we do add support for some fields that can't be extracted from pods and need additional API calls, let's not turn them on by default.

**Comment by chaitanyaphalak** [comment]
Proceeding with 3 PR's for now 

1. For adding semantic conventions to the spec https://github.com/open-telemetry/opentelemetry-specification/pull/1723
2. Changing configs for pod k8s-tagger https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3544
3. This PR for namespace k8s-tagger metadata, labels and annotation

**Comment by dmitryax** [comment]
Looks like there is a failing unit test: `TestPodClient_NamespaceToRunningPodNum`

**Comment by tigrannajaryan** [comment]
@owais @dmitryax PTAL.

File: processor/k8sprocessor/client_test.go
Comment (owais) [inline processor/k8sprocessor/client_test.go:69]: could directly return the results

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:107]: Looks like we will instantiate a namespace informer irrespective of whether a user configured the processor to extract any labels/annotations from namespaces or not. I think we should do it only if the processor actually need to use it.

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:124]: This feels a bit awkward. Why would we run one informer as a go routine but another as a regular function. What implications does it have?  At first glace it looks like this method (Start) was earlier waiting for `c.informer.Run` function call to finish but won't do it after this change. Not sure how it affects the behavior. We should make sure it doesn't change the behavior in any way or at least understand all implications and share them in the PR or as a comment.

In case this does not make any practical difference, can we run both informers as go routines for consistency in code?

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:331]: I'd define the `From` field as a custom string type or at least use constants for the values (namespace, pod)

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:327]: are `NamespaceUID` and `NamespaceStartTime` new rules we added or were they always supported? They don't seem to be very useful TBH unlike a Pod's start time. If they are new (we not present in previous release), I think we should drop them to protect users from unknowingly watching all namespaces in a cluster for seemingly little benefit.

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:490]: Looks like we don't need to use defer here since the logic to be protected is quite simple and there is no risk of escaping a lock with an early return. May be guard the map assignment inside a lock/unlock pair directly?

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:499]: could move unblock to end of function instead of defer

**Review by owais — COMMENTED** [review]
Thanks. Overall direction looks good. Took a quick look and left some comments. Will review again in detail in the morning.

File: processor/k8sprocessor/kube/client.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go:124]: we can run both as goroutines for consistency.

File: processor/k8sprocessor/kube/client.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go:327]: These were added as new metadata extracted from the namespace object. I agree they are not much useful and will drop them(Also we don't need them right now)

File: processor/k8sprocessor/client_test.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/client_test.go:69]: @owais Could you elaborate a little more on this one?
I did the same return as the GetPod function above it for consistency.

Comment (owais) [reply processor/k8sprocessor/client_test.go:69]: It's not a big deal. I mean you could write it as `return f.Namespaces[namespace]`

Comment (chaitanyaphalak) [reply processor/k8sprocessor/client_test.go:69]: @owais I think I will skip this one, but I resolved all the other comments. :)

File: processor/k8sprocessor/kube/client.go
Comment (owais) [reply processor/k8sprocessor/kube/client.go]: Are you sure that is safe? Should the processor wait for the Run function to return before it can tell the pipeline it is ready to receive data or is that not the case?

File: processor/k8sprocessor/client_test.go
Comment (owais) [reply processor/k8sprocessor/client_test.go:69]: OK :)

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:490]: Can you please add a comment here describing why this does not need the grace period support like pods have (with deleteQueue)

File: processor/k8sprocessor/kube/client.go
Comment (owais) [inline processor/k8sprocessor/kube/client.go:188]: if `forgetNamespace` is being called only from here and just removes a key from a map, probably the code can be inlined here directly.

File: processor/k8sprocessor/kube/client.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go:108]: By no-op do you mean a mock informer? Like we use in the tests?

File: processor/k8sprocessor/kube/client.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go:490]: Why do we need the grace period implementation for pods? I didnt see any comments explicitly mentioning why we need it?
If I know that I can add a comment for the namespace stuff too :)

Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go:490]: So I think this wouldnt happen with namespaces. Its because when a namespace is deleted all the pods(and other k8s objects in that namespace) in that namespace are deleted before it. So we wont have any spans that might need namespace annotations and labels. Is this cool? I can add it as a comment if you recommend it.

Comment (chaitanyaphalak) [reply processor/k8sprocessor/kube/client.go]: I added the following
`			// When a namespace is deleted all the pods(and other k8s objects in that namespace) in that namespace are deleted before it.
			// So we wont have any spans that might need namespace annotations and labels.
			// Thats why we dont need an implementation for deleteQueue and gracePeriod for namespaces.`

Comment (owais) [reply processor/k8sprocessor/kube/client.go]: It might still happen if the spans are delayed but very unlikely. comment looks fine.

File: processor/k8sprocessor/config.go
Comment (owais) [inline processor/k8sprocessor/config.go:118]: Should document that it defaults to "pod"

**Review by owais — COMMENTED** [review]
Looks good but I can't find where we are setting default value for "from" to "pod". If we aren't yet, let's add that.

File: processor/k8sprocessor/kube/fake_informer.go
Comment (owais) [inline processor/k8sprocessor/kube/fake_informer.go:152]:
```suggestion
type NoOpController struct{
  hasStopped bool
}

func (c *NoOpController) Run(stopCh <-chan struct{}) {
    go func() {
        <- stopCh
        c.hasStopped = true
    }()
}

 func (c *NoOpController) HasStopped() bool {
     return c.hasStopped
 }

 func (c *NoOpController) HasSynced() bool {
     return true
 }
```

Comment (owais) [reply processor/k8sprocessor/kube/fake_informer.go:152]: May be we should do this for correctness?

Comment (owais) [reply processor/k8sprocessor/kube/fake_informer.go:152]: Just in case something in future relies on HasStopped and does something accordingly.

**Comment by chaitanyaphalak** [comment]
@owais The PR is ready.

File: processor/k8sprocessor/options.go
Comment (owais) [inline processor/k8sprocessor/options.go:147]: 1. We return a helpful error if an invalid value is specified for `From` field like this one: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3384/files#diff-224a054603fb24c9548928fedba75a1600a41b427a1d12df3b88a847e486ce71R210
2. I think the default needs to be set and error needs to be returned outside this check (if name == "").
3. This might read cleaner as a switch case but that's totally up to you.

**Review by owais — CHANGES_REQUESTED** [review]

File: processor/k8sprocessor/options.go
Comment (chaitanyaphalak) [reply processor/k8sprocessor/options.go:147]:
```go
		if name == "" {
			switch a.From {
			case metdataNamespace:
				name = fmt.Sprintf("k8s.namespace.%s.%s", fieldType, a.Key)
				// By default if the From field is not set for labels and annotations we want to extract them from pod
			case "", metadataPodName:
				name = fmt.Sprintf("k8s.pod.%s.%s", fieldType, a.Key)
				a.From = kube.MetadataFromPod
			default:
				return rules, fmt.Errorf("from must contain exactly one value like pod or namespace")
			}
		}
```

Comment (owais) [reply processor/k8sprocessor/options.go:147]: This is fine except `a.From = kube.MetadataFromPod` should be before the `if` condition because we don't know where else `a.From` might be used so we shouldn't limit setting the defaults only when `name == ""`. Plus, in general we shouldn't mix figuring out values for two different fields in the same logical block IMO for clarity.

File: processor/k8sprocessor/config_test.go
Comment (owais) [inline processor/k8sprocessor/config_test.go]: nit: this should be part of the 3rd import group below

File: processor/k8sprocessor/options.go
Comment (owais) [inline processor/k8sprocessor/options.go:146]: nit: "like pod or namespace" suggests it accepts other values that are like these such as deployments. It is also helpful to print out the user specified value so it is easy to spot the error in config in large and complex files. 

I think this should read something like: `"%s" is not a valid choice for `From`. Must be one of: "pod", "namespace"` where `%s` gets replaced with the user specified value.

**Review by owais — APPROVED** [review]
Looks good but left 2 minor comments. I trust you to take care of those before getting it merged :)

Thanks for the contribution. great work!

**Review by tigrannajaryan — COMMENTED** [review]
Please update the documentation in doc.go to show how the new functionality can be used.

**Pull request merged**
